hadoop-hdds/docs/content/concept/Recon.md

```diff
@@ -0,0 +1,166 @@
+`UPDATE`, etc. on either OM db or SCM db. Based on this, a task either implements 
+`org.apache.hadoop.ozone.recon.tasks.ReconOmTask` or extends 
+`org.apache.hadoop.ozone.recon.scm.ReconScmTask`.   
+
+An example `ReconOmTask` is `ContainerKeyMapperTask` that persists the container -> key 
+mapping in RocksDB. This is useful to understand which keys were part of the container 
+when the container is reported missing or is in a bad health state. Another example is 
+`FileSizeCountTask` which keeps track of count of files within a given file size range in 
+a SQL database. These tasks have implementations for two scenarios:
+ 
+ - Full snapshot (reprocess())
+ - Delta updates (process())
+ 
+When a full snapshot of OM db is obtained from the leader OM, the reprocess() 
+is called on all the registered OM tasks. On subsequent delta updates, process()
+is called on these OM tasks. 
+
+An example `ReconScmTask` is `ContainerHealthTask` that runs in configurable 
+intervals to scan the list of all the containers and to persist the state of 
+unhealthy containers (`MISSING`, `MIS_REPLICATED`, `UNDER_REPLICATED`, `OVER_REPLICATED`)
+in a SQL table. This information is used to determine if there are any missing 
+containers in the cluster.
+
+## Recon and Prometheus
+
+Recon can integrate with any Prometheus instance configured to collected metrics 
+and can display useful information in Recon UI in Datanodes and Pipelines pages.
+Recon also exposes a proxy endpoint ([/metrics]({{< ref "interface/ReconApi.md#metrics" >}}))
+to query Prometheus. This integration can be enabled by setting this configuration `ozone.recon.prometheus.http.endpoint` 
+to the Prometheus endpoint like `ozone.recon.prometheus.http.endpoint=localhost:9090`.
+
+## API Reference
+
+[Link to complete API Reference]({{< ref "interface/ReconApi.md" >}})
+   
+## Persisted state
+
+ * A local copy of [OM database]({{< ref "concept/OzoneManager.md#persisted-state" >}})
+ * A local copy of [SCM database]({{< ref "concept/StorageContainerManager.md#persisted-state" >}})
+ * The following data is persisted in Recon in the specified RocksDB directory:
+     * ContainerKey table
+       * Stores the mapping (container, key) -> count
+     * ContainerKeyCount table
+       * Stores containerID -> no. of keys count within the container
+ 
+ * The following data is stored in the configured SQL database (default is Derby):
+     * GlobalStats table
+       * A Key -> Value table to store aggregate information like total
+       number of volumes / buckets / keys present in the cluster
+     * FileCountBySize table
+       * Keeps track of the number of files present within a file size range in the cluster
+     * ReconTaskStatus table
+       * Keeps track of the status and last run timestamp of the registered OM and SCM 
+       db tasks in the [Recon Tasks Framework](#tasks-framework)
+     * ContainerHistory table
```

Comment:
Just FYI, the 'ContainerHistory' SQL table will likely be removed in HDDS-4403.

Reply:
Ack.